FEATURE: add custom localization support

jnelle opened the pull request:
Signed-off-by: Billaids <jimmy.nelle@hsw-stud.de>

Commit:
add custom localization support

Signed-off-by: Billaids <jimmy.nelle@hsw-stud.de>

Commit:
add missing parameter

Signed-off-by: Billaids <jimmy.nelle@hsw-stud.de>

jnelle commented:
![Screenshot 2021-08-17 095741](https://user-images.githubusercontent.com/36324542/129687291-6da2c1fa-373d-4ae0-ab23-3416a239e79e.png)

works fine for me

MayLau-CbL commented:
Thanks for contributing @Billaids! I will review this PR this weekend.

MayLau-CbL commented on lib/date_item_widget.dart:
```diff
@@ -64,7 +71,9 @@ class DateItemWidget extends StatelessWidget {
             switch (this.dateItemComponentList[index]) {
               case DateItem.WeekDay:
-                return Text(DateFormat.E().format(this.dateTime),
+                return Text(
+                    DateFormat.E(Locale(this.locale).toString())
+                        .format(this.dateTime),
```
Which flutter version are you using? I am using 2.2.3 and the it keeps warning me `The argument type 'String?' can't be assigned to the parameter type 'String'.`

jnelle commented:
Hey, I'm using 2.5.0-5.1.pre it's flutter beta

MayLau-CbL commented on lib/date_item_widget.dart:
```diff
@@ -29,12 +30,17 @@ class DateItemWidget extends StatelessWidget {
 
+  // locale String like "de"
+  // https://api.flutter.dev/flutter/date_symbol_data_http_request/availableLocalesForDateFormatting.html
+  final String? locale;
```
```suggestion
  final String locale;
```

MayLau-CbL commented on lib/horizontal_date_picker.dart:
```diff
@@ -69,6 +69,8 @@ class HorizontalDatePickerWidget extends StatefulWidget {
   ///at least one info must be in the list
   final List<DateItem> dateItemComponentList;
 
+  final String locale;
```
```suggestion
  final String? locale;
```

MayLau-CbL commented on lib/horizontal_date_picker.dart:
```diff
@@ -94,6 +96,7 @@ class HorizontalDatePickerWidget extends StatefulWidget {
     this.monthFontSize = 12,
     this.dayFontSize = 18,
     this.weekDayFontSize = 12,
+    required this.locale,
```
```suggestion
   this.locale,
```

MayLau-CbL commented on lib/horizontal_date_picker.dart:
```diff
@@ -159,6 +162,7 @@ class _HorizontalDatePickerWidgetState
               }
             },
             child: DateItemWidget(
+              locale: widget.locale,
```
```suggestion
              locale: widget.locale ?? Intl.systemLocale,
```

MayLau-CbL reviewed (commented):
I think the warning issue is related to beta version changes. Anyway to support both stable/beta version and adapt to previous version, i have made some suggestion on the widget params:
1. HorizontalDatePickerWidget's locale allow user pass null or nothing in order to let existing users have no needs to make any changes
2. make DateItemWidget's locale as String (non-nullable) for stable flutter version Locale class
3. when passing locale from HorizontalDatePickerWidget to DateItemWidget, if HorizontalDatePickerWidget's locale is null, will default pass the system locale to DateItemWidget

jnelle committed:
Update lib/horizontal_date_picker.dart

what's the return from Intl.systemLocale? The return value should be like here in this list

https://api.flutter.dev/flutter/date_symbol_data_http_request/availableLocalesForDateFormatting.html

Co-authored-by: May Lau <may.lau.dev@gmail.com>

jnelle commented:
what's the return from Intl.systemLocale? The return value should be like here in this list
https://api.flutter.dev/flutter/date_symbol_data_http_request/availableLocalesForDateFormatting.html

MayLau-CbL commented:
> what's the return from Intl.systemLocale? The return value should be like here in this list
> https://api.flutter.dev/flutter/date_symbol_data_http_request/availableLocalesForDateFormatting.html

normally should be en_US, depends on the system setting at that time. Or for more secure, we can set it as` locale: widget.locale ?? 'en_US';` directly as it is only a default setting for this widget

jnelle commented:
> > what's the return from Intl.systemLocale? The return value should be like here in this list
> > https://api.flutter.dev/flutter/date_symbol_data_http_request/availableLocalesForDateFormatting.html
> 
> normally should be en_US, depends on the system setting at that time. Or for more secure, we can set it as` locale: widget.locale ?? 'en_US';` directly as it is only a default setting for this widget

Yes, this is a good idea!

jnelle commented:
Could you merge it? :D

MayLau-CbL commented:
sorry for the later reply, i dunt feel so well recently.🤦‍♀️ do you mind applying those comments to the branch first if you have no question on them? if ur branch is not convenient for merging those changes, i can fork yours and apply them later.
